core/src/agents/base_agent.ts

```diff
@@ -124,13 +125,14 @@ export abstract class BaseAgent {
   async *
       runAsync(parentContext: InvocationContext):
           AsyncGenerator<Event, void, void> {
-    const span = trace.getTracer('gcp.vertex.agent')
-                     .startSpan(`agent_run [${this.name}]`);
-    try {
-      const context = this.createInvocationContext(parentContext);
-
+    const span = tracer.startSpan(`invoke_agent ${this.name}`);
+    const ctx = trace.setSpan(context.active(), span);
+    const self = this;
+    const gen = asyncGeneratorWithContext<Event, void, undefined>(ctx, async function* () {
```

Comment:
Can you try another solution for this problem: 
https://github.com/open-telemetry/opentelemetry-js/issues/2951#issuecomment-1214587378

In that case you can just bind `this` to generator function and avoid defining `self` object to keep the context.
It will reduce number of changes lines and will improve readability. 

```suggestion
const gen = bindAsyncGenerator<Event, void, undefined>(this, async function* () {
....
});